### DISCUSSION TITLE
Home-Assistant zwingend erforderlich?

### DISCUSSION THREAD

**Discussion by d4n7e1** [s-allius/tsun-gen3-proxy#59] [Q&A] [answered]
Guten Tag und vielen Dank für die tolle Idee und Umsetzung dieses Projekts!

Aktuell habe ich leider Probleme, den tsun-proxy ans Laufen zu bekommen. Da ich kein HA nutze und bereits einen MQTT-Broker am Laufen habe, habe ich den tsun-proxy via Stack und Portainer deployed.

Der erste Start funktionierte Problemlos und im log hat er lediglich moniert, dass kein MQTT-Broker konfiguriert sei. Nachdem ich dann aber die config.toml angepasst habe, erhalte ich bei jedem Containerstart folgenden Fehler:

```
Traceback (most recent call last):

  File "/home/tsun-proxy/./server.py", line 85, in <module>

    Inverter.class_init()

  File "/home/tsun-proxy/inverter.py", line 21, in class_init

    cls.entity_prfx = ha['entity_prefix'] + '/'

                      ~~^^^^^^^^^^^^^^^^^

KeyError: 'entity_prefix'
```

Meine config.toml habe ich nicht nennenswert verändert, lediglich die für mich erforderlichen Anpassungen vorgenommen (SERIENNUMMER ersetzt meine tatsächliche S/N):
```
# configuration to reach tsun cloud
tsun.enabled = true   # false: disables connecting to the tsun cloud, and avoids updates
tsun.host    = 'logger.talent-monitoring.com'
tsun.port    = 5005

# configuration to reach the new tsun cloud for G3 Plus inverters
#solarman.enabled = true   # false: disables connecting to the tsun cloud, and avoids updates
#solarman.host    = 'iot.talent-monitoring.com'
#solarman.port    = 10000

# mqtt broker configuration
mqtt.host    = '192.168.14.98'   # URL or IP address of the mqtt broker
mqtt.port    = 1883
mqtt.user    = ''
mqtt.passwd  = ''


# home-assistant
ha.auto_conf_prefix = 'homeassistant'       # MQTT prefix for subscribing for homeassistant status updates
ha.discovery_prefix = 'homeassistant'       # MQTT prefix for discovery topic 
ha.entity_prefix    = 'tsun'                # MQTT topic prefix for publishing inverter values
ha.proxy_node_id    = 'proxy'               # MQTT node id, for the proxy_node_id
ha.proxy_unique_id  = 'P17SERIENNUMMER1'    # MQTT unique id, to identify a proxy instance

# microinverters
inverters.allow_all = true   # allow inverters, even if we have no inverter mapping

# inverter mapping, maps a `serial_no* to a `mqtt_id` and defines an optional `suggested_place` for `home-assistant`
#
# for each inverter add a block starting with [inverters."<16-digit serial numbeer>"]
[inverters."R17SERIENNUMMER1"]
node_id = 'wechselrichter'                 # Optional, MQTT replacement for inverters serial number  
#suggested_area = ''          # Optional, suggested installation area for home-assistant
#pv1 = {type = 'RSM40-8-395M', manufacturer = 'Risen'}   # Optional, PV module descr
#pv2 = {type = 'RSM40-8-395M', manufacturer = 'Risen'}   # Optional, PV module descr

#[inverters."R17xxxxxxxxxxxx2"]
#node_id = ''                 # Optional, MQTT replacement for inverters serial number  
#suggested_area = ''          # Optional, suggested installation area for home-assistant
#pv1 = {type = 'RSM40-8-405M', manufacturer = 'Risen'}   # Optional, PV module descr
#pv2 = {type = 'RSM40-8-405M', manufacturer = 'Risen'}   # Optional, PV module descr

#[inverters."Y170000000000001"]
#monitor_sn = 2000000000       # The "Monitoring SN:" can be found on a sticker enclosed with the inverter
#node_id = ''                 # Optional, MQTT replacement for inverters serial number  
#suggested_area = ''          # Optional, suggested installation place for home-assistant
#pv1 = {type = 'RSM40-8-410M', manufacturer = 'Risen'}   # Optional, PV module descr
#pv2 = {type = 'RSM40-8-410M', manufacturer = 'Risen'}   # Optional, PV module descr
#pv3 = {type = 'RSM40-8-410M', manufacturer = 'Risen'}   # Optional, PV module descr
#pv4 = {type = 'RSM40-8-410M', manufacturer = 'Risen'}   # Optional, PV module descr
```
Ich habe auch bereits versucht, den "ha." Bereich auszukommentieren, ohne Erfolg. Entferne ich beim mqtt-broker für die IP die '', so erhalte ich wieder die Meldung der fehlenden MQTT-Konfiguration wie beim ersten Start.

Aufgrund der Fehlermeldung habe ich die Befürchtung, dass ein laufender HA erforderlich ist, mit welchem ich bisher noch keinerlei Erfahrung habe, da ich alles über openhab verwalte. Vielleicht habe ich aber auch nur an anderer Stelle einen Denkfehler?

Beste Grüße

Arne

**Comment by s-allius** [comment, marked as answer]
Hallo Arne,

schön das du dich für die Lösung interessierst. Prinzipiell kann man den Proxy auch ohne HA nutzen. Allerdings ist der Parser für die config.toml Datei recht einfach gehalten und die Mandantory Parameter müssen vorhanden sein. Fehlt ein Parameter oder es gibt einen Syntaxfehler in der Datei, dann bricht der Parser ab und es kommt zu den von dir dargestellten Folgefehler, da dann intern eine leere Konfiguration vorliegt.
Ich vermute, dass du noch die Version 0.6.0 verwendest, da ich in der 0.7.0 da schon etwas geändert habe. 
Ich schaue mir jetzt mal deine toml Datei genauer an, um dir konkretere Hilfe zu geben.

Update
---
Also die Zeilen mit den Paramtern `tsun`, `solarman`, `mqtt`und `ha`sind alle mandantory und müssen gültig sein. Du hast einige auskommentiert, was vermutlich zum Problem führt.

Ich werde in späteren Version einen Schalter zum Deaktivieren des HA-Autoconfig Features einbauen, sodass diese Topics dann nicht mehr per MQTT propagiert werden. Solange musst du leider die Parameter spezifizieren.

Gruß Stefan

**Reply by d4n7e1** [reply]
Guten Abend Stefan,

Gutes kann so einfach sein! In meinem Fall war es tatsächlich das solarman, welches nicht auskommentiert sein durfte. Nach entfernen des # startete der Container fehlerfrei und mein Broker erhält die Daten :-) Ich bin fälschlicherweise davon ausgegangen, dass ich diesen Bereich nicht benötige, da ich ein MS600 Gen3 habe und somit die Einträge aus dem tsun-Bereich relevant sind. 

Jetzt bin ich mal gespannt, was da alles an Werten die nächsten Tage hineintrudelt. Vielen Dank für die schnelle Hilfe und dass du dein Können in solch ein tolles Projekt fließen lässt um die Daten mit den eigenen Systemen zu erfassen!

Beste Grüße

Arne
